### PR TITLE
coord: Dump the LastMessage to stderr if the Coordinator panics

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds a Drop impl to the `LastMessage` struct that we use for slow message tracking, to also print stderr if we're panicking while getting dropped.

### Motivation

When the Coordinator panics it is very likely that we will not record the last run statement as part of statement logging. This is a last ditch effort to capture debug info about what the Coordinator was doing when we panicked.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a, internal debugging change

### CODE REVIEW

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -2573,6 +2568,23 @@ impl Coordinator {
...
+    kind: &'static str,
+    stmt: Option<Arc<Statement<Raw>>>,
+}
+
+impl Drop for LastMessage {
+    fn drop(&mut self) {
+        // Only print the last message if we're currently panicking, otherwise we'd spam our logs.
+        if std::thread::panicking() {
+            // If we're panicking theres no guarantee `tracing` still works, so print to stderr.
+            eprintln!("Coordinator panicking, dumping last message\n{:?}", self);
```
Comment (ParkMyCar) [inline src/adapter/src/coord.rs:2583]: using `tracing::error!` here would be preferred but I don't think we're guaranteed to have tracing work. But I'm also not sure how easy it is to read stderr?

Comment (benesch) [reply src/adapter/src/coord.rs:2583]: We should do the same thing we do for logging in the stuck coordinator case and redact the SQL, to keep customer data out of logs.

**Review by benesch — APPROVED** [review]
This is a great idea! I'm not sure exactly how well this will work in practice given that we do various things off the main coordinator thread now, and so the last message received may not actually be the one that caused the panic. But it is a strict improvement over the current situation.

File: src/adapter/src/coord.rs (hunk 1 above)
Comment (benesch) [reply src/adapter/src/coord.rs:2583]: But yeah, agreed on `eprintln!` rather than `tracing::error!`. I think it should go to Loki just fine.

Comment (ParkMyCar) [reply src/adapter/src/coord.rs]: Good call! Updated to use the redacted SQL

**Pull request merged**